plugins/inputs/sqlserver/sqlserver.go

```diff
@@ -272,15 +280,48 @@ func (s *SQLServer) Start(acc telegraf.Accumulator) error {
 		return err
 	}
 
-	if len(s.Servers) == 0 {
-		s.Servers = append(s.Servers, defaultServer)
-	}
+	// initialize mutual exclusion lock
+	muCacheLock = sync.RWMutex{}
 
 	for _, serv := range s.Servers {
-		pool, err := sql.Open("mssql", serv)
-		if err != nil {
-			acc.AddError(err)
-			return err
+		var pool *sql.DB
+
+		// setup connection based on authentication
+		rx := regexp.MustCompile(`\b(?:(Password=((?:&(?:[a-z]+|#[0-9]+);|[^;]){0,})))\b`)
+
+		// when password is provided in connection string, use SQL auth
+		if rx.MatchString(serv) {
+			var err error
+			pool, err = sql.Open("mssql", serv)
+
+			if err != nil {
+				acc.AddError(err)
+				return err
+			}
+		} else {
+			// otherwise assume AAD Auth with system-assigned managed identity (MSI)
+
+			// AAD Auth is only supported for Azure SQL Database or Azure SQL Managed Instance
+			if s.DatabaseType == "SQLServer" {
+				return fmt.Errorf("Database connection failed : AAD auth is not supported for SQL VM i.e. DatabaseType=SQLServer")
+			}
+
+			// get token from im-memory cache variable or from Azure Active Directory
+			tokenProvider, err := getTokenProvider()
+			if err != nil {
+				fmt.Printf("Error creating AAD token provider for system assigned Azure managed identity : %s", err.Error())
```

Comment:
Never `fmt.Printf`! Add a logger facility 
```
	Log telegraf.Logger `toml:"-"`
```
to the `SQLServer` struct and then use `s.Log.Warn()` or whatever to log the message. Alternatively you can just return the message as error message!?

Reply:
Yes, I noticed it after merging the latest changes (for connection pooling). I was already adding the error to accumulator and this printf is not required, refactored it.